src/credentials.rs

```diff
@@ -0,0 +1,105 @@
+use crate::errors::SdError;
+use nix::dir;
+use nix::fcntl::OFlag;
+use nix::sys::stat::Mode;
+use std::env;
+use std::fs::File;
+use std::path::PathBuf;
+
+/// Credential loader for units.
+///
+/// Credentials are read by systemd on unit startup and exported by their ID.
+///
+/// **Note**: only the user associated with the unit and the superuser may access credentials.
+///
+/// More documentation: <https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Credentials>
+#[derive(Debug)]
+pub struct CredentialsLoader {
+    path: PathBuf,
+    _dirfd: dir::Dir,
+}
+
+impl CredentialsLoader {
+    /// Try to open credentials directory.
+    pub fn open() -> Result<Self, SdError> {
```

Comment:
Shouldn't we expose the actual IOError here?  

As far as I can see all errors in this function are IOErrors. I feel that swallowing these and returning our own error type with a generic message would loose some valuable information for debugging 🤔

Reply:
Also answering all the error questions below here.

I'd say no, for a few reasons.
The main one is that I don't really want to have people matching stuff like `NotFound` here. Probing is meant to be performed through `path_from_env()`, and a `Some` there can only correspond to an `Ok` here. Perhaps that method should have a different name, and the overall module better documentation?

The second one is that I prefer consumers to have a consistent `SdError` type everywhere, with verbose details about what happened. This is more useful to applications than a generic `No such file or directory`: I already had to debug too many cases where there was some issues but the error message was not carrying enough details :wink:

Reply:
> The main one is that I don't really want to have people matching stuff like NotFound here. 

I wouldn't want people to match on these either, but I'd like to give callers the ability to thoroughly debug issues, and from my point of view this includes being able to dump full details of errors. 

I'm fine with wrapping errors up into our own type, and I like a consistent error type too, but for debugging I'd also like to have access to the internal error, as a principle.

But I also understand that there's not much that can go wrong here, and we retain the original error message, so it's still not my preferred API, but it's okay, and in the end it's your call 🙂 

> The second one is that I prefer consumers to have a consistent SdError type everywhere, with verbose details about what happened. This is more useful to applications than a generic No such file or directory: I already had to debug too many cases where there was some issues but the error message was not carrying enough details wink

I feel your pain; that's why I'm a fan of stacking errors. I love the context type of `anyhow`, and that's why I'd love to have stacked errors with e.g. `thiserror` here as well.

But as said, I'm okay with this API 🙂